### DISCUSSION TITLE
Best approach for deeply nested JSON

### DISCUSSION THREAD

**Discussion by gotdan** [duckdb/duckdb#5272] [Q&A] [answered]
Hi, I have a nested json structure that I want to flatten in duckdb by selecting one or more objects at each level based on their properties to walk down the hierarchy, and then returning a property from each of the lowest level objects as rows in the output. The number of levels is fixed, but the number and structure of the objects at each level varies. I've tacked this by using the UNNEST(from_json()) approach I found on another thread and a bunch of CTEs. Below is a simplified, self-contained example of what I'm trying. 

Is this the best way to do this in duckdb (in postgres, I'd probably do left joins to json_array_elements expressions)? Is it likely perform very slowly on large datasets? Thanks!

```sql
-- in the real db, this would be read from an ndjson file or loaded into a table
WITH data AS (
	SELECT '{
		"id": "object-1",
		"children_first": [{
			"type": "deep_1",
			"children_second": [{
				"type": "deep_1_2",
				"children_third": [{
					"type": "deep_1_2_3",
					"id": "leaf_1",
					"value": 123
				},{
					"type": "deep_1_2_3",
					"id": "leaf_2",
					"value": 456
				}]
			}]
		}]
	}'::JSON AS raw
),
first_level AS (
	SELECT
		raw ->> 'id' AS id,
		UNNEST(FROM_JSON(raw ->> 'children_first', '["JSON"]')) AS children_1
	FROM data
),
second_level AS (
	SELECT 
		id,
		UNNEST(FROM_JSON(children_1 ->> 'children_second', '["JSON"]')) AS children_2
	FROM first_level 
	-- not required in the example, but need the option to filter children at each level by an object property
	WHERE (children_1 ->> 'type') = 'deep_1'
),
third_level AS (
	SELECT 
		id,
		UNNEST(FROM_JSON( children_2 ->> 'children_third', '["JSON"]')) AS children_3
	FROM second_level
	-- not required in the example, but need the option to filter children at each level by an object property
	WHERE (children_2 ->> 'type') = 'deep_1_2'
)

SELECT
	id,
	children_3 ->> 'id' AS leaf_id,
	children_3 ->> 'value' AS leaf_value
FROM third_level
```

**Comment by lnkuiper** [comment, marked as answer]
Minor detail, the `->` operator returns data of type `JSON`, whereas the `->>` operator returns data of type `VARCHAR`, just like in postgres. The query above will work, as the data will be implicitly casted back to `JSON` to use in the `FROM_JSON` function,  however, there is a subtle difference in other cases, e.g.:
```sql
D select '{"duck":"goose"}'->'duck';
┌──────────────────────────────┐
│ '{"duck":"goose"}' -> 'duck' │
├──────────────────────────────┤
│ "goose"                      │
└──────────────────────────────┘
-- ^ the result is still JSON, hence the quotes around goose
D select '{"duck":"goose"}'->>'duck';
┌─────────────────────────────────┐
│ ('{"duck":"goose"}' ->> 'duck') │
├─────────────────────────────────┤
│ goose                           │
└─────────────────────────────────┘
-- ^ the result is VARCHAR, no quotes
```

That being said, the above query seems like a reasonable approach to deal with deeply nested JSON. However, every time you manipulate JSON with the `FROM_JSON` and `->` or `->>` functions/operators, the JSON will be re-parsed. This causes a lot of redundant work.

You can consider doing a single `FROM_JSON` on your data. You can get the structure of the JSON in your table like so:
```sql
SELECT json_structure(raw) FROM data LIMIT 1;
```

This gives you:
```json
{
  "id":"VARCHAR",
  "children_first":[
    {
      "type":"VARCHAR",
      "children_second":[
        {
          "type":"VARCHAR",
          "children_third":[
            {
              "type":"VARCHAR",
              "id":"VARCHAR",
              "value":"UBIGINT"
            }
          ]
        }
      ]
    }
  ]
}
```

You can now use the given structure to parse the entirety of your data once and create DuckDB nested types:
```sql
select from_json(raw, '{"id":"VARCHAR","children_first":[{"type":"VARCHAR","children_second":[{"type":"VARCHAR","children_third":[{"type":"VARCHAR","id":"VARCHAR","value":"UBIGINT"}]}]}]}') from test;
```

This will give you back a column of type:
```
STRUCT(id VARCHAR, children_first STRUCT(type VARCHAR, children_second STRUCT(type VARCHAR, children_third STRUCT(type VARCHAR, id VARCHAR, value UBIGINT)[])[])[])
```
Accessing the data within our native list/struct types is much, much more efficient than parsing JSON.

Putting it all together, you can re-write your query like so:
```sql
WITH data AS (
	SELECT '{
		"id": "object-1",
		"children_first": [{
			"type": "deep_1",
			"children_second": [{
				"type": "deep_1_2",
				"children_third": [{
					"type": "deep_1_2_3",
					"id": "leaf_1",
					"value": 123
				},{
					"type": "deep_1_2_3",
					"id": "leaf_2",
					"value": 456
				}]
			}]
		}]
	}'::JSON AS raw
), shredded_data AS (
	SELECT from_json(raw, '{"id":"VARCHAR","children_first":[{"type":"VARCHAR","children_second":[{"type":"VARCHAR","children_third":[{"type":"VARCHAR","id":"VARCHAR","value":"UBIGINT"}]}]}]}') not_raw
	FROM data
), first_level AS (
	SELECT not_raw.id,
		   UNNEST(not_raw.children_first) AS children_1
	FROM shredded_data
), second_level AS (
	SELECT id,
	       UNNEST(children_1.children_second) AS children_2
    FROM first_level
    WHERE children_1.type = 'deep_1'
), third_level AS (
	SELECT id,
	       UNNEST(children_2.children_third) AS children_3
    FROM second_level
    WHERE children_2.type = 'deep_1_2'
)
SELECT id,
	   children_3.id AS leaf_id,
	   children_3.value AS leaf_value
FROM third_level;
```

Which will give you:
```
┌──────────┬─────────┬────────────┐
│    id    │ leaf_id │ leaf_value │
├──────────┼─────────┼────────────┤
│ object-1 │ leaf_1  │ 123        │
│ object-1 │ leaf_2  │ 456        │
└──────────┴─────────┴────────────┘
```

This should be much more performant. This query works for me, but I have to give a disclaimer: I haven't tested `FROM_JSON` on large amounts of such deeply nested JSON. If you run into any issues, I would love to hear about them, and I will try to fix them.

**Reply by cameronneylon** [reply]
I'm coming to a very similar issue in 2025. This solution works perfectly well. I am wondering whether doing this with a list_filter or list_extract approach would be expected to be more performant now (I think these may not have been available when the answer was originally written)?

**Reply by lnkuiper** [reply]
`list_extract` and `list_filter` do different things, I'm not sure it makes sense to compare them based on performance.

If you want to get a single list element by index, use `list_extract`.
If you want to get multiple list elements based on some kind of filter, use `list_filter`.

**Reply by cameronneylon** [reply]
Apologies, I wasn't clear. I meant to compare the UNNEST approach to something similar using list_filter but that isn't really the nature of the OP question so I'm just confusing the issue. I will ask my specific question separately with some example data! Thanks!

**Comment by gotdan** [comment]
Thank you @lnkuiper! Your explanation is very helpful and the re-written query really clarifies how from_json works for me - parsing the json once seems like a definite improvement!

I have a related question after playing around more with the query. It seems like the unnest function acts like an implicit inner join, so if the json data has an `id` property, but the `children_first` property isn't populated, no rows will be returned. Is there a way to make the unnest conditional on the property being populated, so it's more like a left join? I suppose another way to address this would be to left join from the shredded_data CTE to the third_level CTE in the query, but not sure if this would be an efficient approach:

```sql
SELECT shredded_data.not_raw.id,
  children_3.id AS leaf_id,
  children_3.value AS leaf_value
FROM shredded_data
LEFT JOIN third_level
  ON shredded_data.not_raw.id = third_level.id
```

**Reply by lnkuiper** [reply]
This is a bit hacky but if you want `NULL` when the list length is 0, you can do this:
```sql
create table test (id int, my_list int[]);
insert into test values (1, [1, 2, 3]), (2, []);
select id, length(my_list) list_len, case when list_len = 0 then unnest([null]) else unnest(my_list) end vals from test;
┌───────┬──────────┬───────┐
│  id   │ list_len │ vals  │
│ int32 │  int64   │ int32 │
├───────┼──────────┼───────┤
│     1 │        3 │     1 │
│     1 │        3 │     2 │
│     1 │        3 │     3 │
│     2 │        0 │       │
├───────┴──────────┴───────┤
│ 4 rows         3 columns │
└──────────────────────────┘
```

I'm not sure how the performance of this compares to the left join approach.

This is also a fun way of doing things without `UNNEST`:
```sql
WITH data AS (
    SELECT '{
        "id": "object-1",
        "children_first": [{
            "type": "deep_1",
            "children_second": [{
                "type": "deep_1_2",
                "children_third": [{
                    "type": "deep_1_2_3",
                    "id": "leaf_1",
                    "value": 123
                },{
                    "type": "deep_1_2_3",
                    "id": "leaf_2",
                    "value": 456
                }]
            }]
        }]
    }'::JSON AS raw
), shredded_data AS (
    SELECT from_json(raw, '{"id":"VARCHAR","children_first":[{"type":"VARCHAR","children_second":[{"type":"VARCHAR","children_third":[{"type":"VARCHAR","id":"VARCHAR","value":"UBIGINT"}]}]}]}') not_raw
    FROM data
)
SELECT not_raw.id,
       flatten(
           flatten(
               [
                   [
                       [
                           {id:c3.id, value:c3.value}
                           for c3 in c2.children_third if c3.type = 'deep_1_2_3'
                       ]
                        for c2 in c1.children_second if c2.type = 'deep_1_2'
                    ]
                    for c1 in not_raw.children_first if c1.type = 'deep_1'
                ]
           )
       ) children_3
FROM shredded_data;
┌──────────┬──────────────────────────────────────────────────────────────┐
│    id    │                          children_3                          │
│ varchar  │             struct(id varchar, value ubigint)[]              │
├──────────┼──────────────────────────────────────────────────────────────┤
│ object-1 │ [{'id': leaf_1, 'value': 123}, {'id': leaf_2, 'value': 456}] │
└──────────┴──────────────────────────────────────────────────────────────┘
```

This uses our python-style list comprehension syntax (currently on master, will be in the next release). It will give you an empty list if the top-level list was empty.

**Comment by gotdan** [comment]
Neat idea, thanks! Once v5 is out, I'll do a performance comparison of the two approaches with real data and report back.

For the longer term, any chance the ability to have a join to an array via [json_array_elements()](https://www.postgresql.org/docs/15/functions-json.html#FUNCTIONS-JSON-PROCESSING-TABLE) and an equivalent for the List type is on the roadmap? It would be great to be more compatible with Postgres and, at least for my use case, I think the SQL would be easier for people to read.

**Reply by lnkuiper** [reply]
We would like to add this feature, but our table-producing functions cannot yet reference columns. This is essentially a lateral join. It is on the roadmap.

**Reply by gotdan** [reply]
Sounds good, thanks!
